### PR TITLE
ADR-0009: Pipeline Service via Operator

### PR DESCRIPTION
**Pull request by Roming22**
- Deploy "Pipeline Service" using the stock OpenShift Pipelines operator, ensuring that the "latest" code is deployed and that all configuration is driven by the operator.
- Mark ADR-0001 as obsolete, superceded by this decision.
- Require disabling of Pipelines pruner in favor of Results-driven pruning.
- Add requirements for hot fixing of operators and controlling operator versions.

This is a rework of #26. Most of those changes have already been implemented, so this PR to catch up with the current design.

### CODE REVIEW

**Review by chmouel — COMMENTED** [review]
sounds good to me w.r.t pac!

SprayProxy is not mentioned anymore in the document but still in the diagram, is that in purpose?

**Comment by Roming22** [comment]
@chmouel Good catch. I've added a small blurb on the subject in the ADR.

File: ADR/0009-pipeline-service-via-operator.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  end users do not need access to the underlying compute cluster(s).
+- The Pipelines as Code application name must match the GitHub Application name.
+- The GitHub Application secret.
+- Any configuration related to performance.
+
+Furthermore, as the service will be accessed through CodeReadyToolchain (CRT), the
+following changes are also specific to RHTAP:
+- Deploying a proxy (known as `SprayProxy`) on the CRT host cluster that redirects
+  incoming PaC requests to the member clusters.
+- Provinding a plugin to the CRT Proxy so Tetkon Result requests are redirected
```
Comment (chmouel) [inline ADR/0009-pipeline-service-via-operator.md:47]: nit
```suggestion
- Providing a plugin to the CRT Proxy so Tetkon Result requests are redirected
```

File: ADR/0009-pipeline-service-via-operator.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+- Direct Pipelines as Code to use a URL pattern that displays the `PipelineRun`
+  or `TaskRun` info from the Hybrid Application Console (HAC). This ensures
+  end users do not need access to the underlying compute cluster(s).
+- The Pipelines as Code application name must match the GitHub Application name.
+- The GitHub Application secret.
+- Any configuration related to performance.
+
+Furthermore, as the service will be accessed through CodeReadyToolchain (CRT), the
+following changes are also specific to RHTAP:
+- Deploying a proxy (known as `SprayProxy`) on the CRT host cluster that redirects
```
Comment (chmouel) [inline ADR/0009-pipeline-service-via-operator.md:45]: i think there is some sort of filtering (or anti ddos/abus mechanism?) in sprayproxy you may want to mention in there...

File: ADR/0009-pipeline-service-via-operator.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+repository, which is then imported into
+[infra-deployments](https://github.com/redhat-appstudio/infra-deployments) as
+an ArgoCD application.
+
+Configurations that are specific to RHTAP must be made available through
+the OSP CRDs. The following changes are specific to RHTAP:
+
+- Disable Tekton Triggers. Pipelines as Code will be the favored mechanism for
+  event-based triggering of pipelines for now. This decision can be revisited
+  in the future based on need.
```
Comment (gabemontero) [inline ADR/0009-pipeline-service-via-operator.md:32]: we are not yet doing this 

this is as much a "generic how you maintain ADRs" question, but I'll ask anway, as I see this in TEPs, KEPs ... I know there is the "proposed" status, and presumably when everything is done, the status becomes "implemented", but shouldn't we provide some indication of what bits are done and what bits are pending?

Comment (gabemontero) [reply ADR/0009-pipeline-service-via-operator.md:32]: I believe this question applies to other bullets here, but I'll refrain from making updates to each line

File: ADR/0009-pipeline-service-via-operator.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+- Disable the pruner that comes with the operator. Tekton Results will be used
+  to prune `PipelineRun` and `TaskRun` data off cluster, thereby ensuring data
+  is archived before removal.
+- Direct Pipelines as Code to use a URL pattern that displays the `PipelineRun`
+  or `TaskRun` info from the Hybrid Application Console (HAC). This ensures
+  end users do not need access to the underlying compute cluster(s).
+- The Pipelines as Code application name must match the GitHub Application name.
+- The GitHub Application secret.
```
Comment (gabemontero) [inline ADR/0009-pipeline-service-via-operator.md:40]: feel like you didn't complete your thought here wrt the GitHub Application Secret

File: book/pipeline-service.md [hunk 5]
```diff
@@ -1,28 +1,50 @@
...
+- Archiving of `PipelineRuns`, `TaskRuns`, and associated logs through Tekton
+  Results.
+
+Pipeline Service also exposes the following ingress points:
+
+- Pipelines as Code Webhook: this is a `Route` that receives webhook events
+  from source code repositories.
+- Tekton Results REST: this is an `Ingress` that serves Tekton Results data
+  over a RESTful API. Clients authenticate with the same `Bearer` token used to
+  authenticate Kubernetes requests.
```
Comment (gabemontero) [inline book/pipeline-service.md:26]: I feel like we should highlight the exporter and that we are providing additional metrics outside of what we currently find upstream, with also noting we are also submitting metrics directly upstream, as well as considering migrating our more complicated downstream metrics upstream once they have matured and solidified.

**Review by gabemontero — COMMENTED** [review]
A couple of questions/suggestions but overall really good job of getting our ADR up to date @Roming22  - thanks

File: ADR/0009-pipeline-service-via-operator.md (hunk 4 above)
Comment (enarha) [reply ADR/0009-pipeline-service-via-operator.md:40]: So this is one of "The following changes are specific to RHTAP:". So the secret is specific/required to/for RHTAP

File: ADR/0009-pipeline-service-via-operator.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+- Any configuration related to performance.
+
+Furthermore, as the service will be accessed through CodeReadyToolchain (CRT), the
+following changes are also specific to RHTAP:
+- Deploying a proxy (known as `SprayProxy`) on the CRT host cluster that redirects
+  incoming PaC requests to the member clusters. Requests are redirected only if:
+  - the webhook secret is valid;
```
Comment (enarha) [inline ADR/0009-pipeline-service-via-operator.md:47]: I suggest "the payload is validated using the webhook secret"

File: ADR/0009-pipeline-service-via-operator.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  - the request size is under 25MB.
```
Comment (enarha) [inline ADR/0009-pipeline-service-via-operator.md:48]: "the payload size ..."

File: ADR/0009-pipeline-service-via-operator.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+Proposed
+
+## Context
+
+kcp is no longer being used as a control plane for RHTAP. This means that
+"Pipeline Service" cannot be deployed as an independent service. For our
+initial MVP, all Tekton APIs need to be deployed onto a standard OpenShift
+cluster (specifically OpenShift Dedicated).
+
+## Decision
```
Comment (enarha) [inline ADR/0009-pipeline-service-via-operator.md:17]: Maybe plural "Decisions" as it looks like they are more than one.

File: ADR/0009-pipeline-service-via-operator.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+- The GitHub Application secret.
+- Any configuration related to performance.
+
+Furthermore, as the service will be accessed through CodeReadyToolchain (CRT), the
+following changes are also specific to RHTAP:
+- Deploying a proxy (known as `SprayProxy`) on the CRT host cluster that redirects
+  incoming PaC requests to the member clusters. Requests are redirected only if:
+  - the webhook secret is valid;
+  - the request size is under 25MB.
+- Providing a plugin to the CRT Proxy so Tetkon Result requests are redirected
```
Comment (enarha) [inline ADR/0009-pipeline-service-via-operator.md:49]: typo "Tekton" and missing S in Results.

File: ADR/0009-pipeline-service-via-operator.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  this setting if `Results` is deployed and configured to prune resources.
+- Pipelines as Code should use an appropriate URL to HAC when interacting with
+  SCM services, such as the GitHub
+  [Checks API](https://docs.github.com/en/rest/guides/getting-started-with-the-checks-api?apiVersion=2022-11-28).
+- Changes to Pipeline components need to be baked into the operator and built
+  rapidly.
+- Hot fixes need to be provided on a "roll forward" basis. OLM Operators do not
+  support "rollback" mechanisms today. Mean time to revert an offending change,
+  rebuild, and deploy needs to be measured in hours.
+- The version of the deployed operator needs to be configurable via ArgoCD.
```
Comment (enarha) [inline ADR/0009-pipeline-service-via-operator.md:68]: We only change the catalog source, specifically the image. The new image is expected to include a newer version of the OSP operator. We also update the ImageContentSourcePolicy where the images required by the operator are pulled from.

File: book/pipeline-service.md [hunk 11]
```diff
@@ -1,28 +1,50 @@
...
+Pipeline Service provides the following:
+
+- Tekton APIs directly through its custom resource definitions.
+- Container image signing and provenance attestations through Tekton Chains.
+- Archiving of `PipelineRuns`, `TaskRuns`, and associated logs through Tekton
+  Results.
+
+Pipeline Service also exposes the following ingress points:
+
+- Pipelines as Code Webhook: this is a `Route` that receives webhook events
```
Comment (enarha) [inline book/pipeline-service.md:22]: Pipelines as Code controller. (we expose the controller and that's also the name of the route pipelines-as-code-controller)

File: book/pipeline-service.md [hunk 12]
```diff
@@ -1,28 +1,50 @@
...
+  from source code repositories.
+- Tekton Results REST: this is an `Ingress` that serves Tekton Results data
```
Comment (enarha) [inline book/pipeline-service.md:24]: Tekton Results API. (we actually support both REST and gRPC)

File: ADR/0009-pipeline-service-via-operator.md (hunk 3 above)
Comment (Roming22) [reply ADR/0009-pipeline-service-via-operator.md:32]: My feeling is that the ADR should capture the destination, and the "current" state should be documented in the service itself.

File: book/pipeline-service.md (hunk 12 above)
Comment (Roming22) [reply book/pipeline-service.md:24]: Updated

File: book/pipeline-service.md (hunk 11 above)
Comment (Roming22) [reply book/pipeline-service.md:22]: Updated

File: ADR/0009-pipeline-service-via-operator.md (hunk 9 above)
Comment (Roming22) [reply ADR/0009-pipeline-service-via-operator.md:49]: Updated

File: ADR/0009-pipeline-service-via-operator.md (hunk 7 above)
Comment (Roming22) [reply ADR/0009-pipeline-service-via-operator.md:48]: Updated

File: ADR/0009-pipeline-service-via-operator.md (hunk 6 above)
Comment (Roming22) [reply ADR/0009-pipeline-service-via-operator.md:47]: Updated

File: ADR/0009-pipeline-service-via-operator.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+- Disable Tekton Triggers. Pipelines as Code will be the favored mechanism for
+  event-based triggering of pipelines for now. This decision can be revisited
+  in the future based on need.
+- Disable the pruner that comes with the operator. Tekton Results will be used
+  to prune `PipelineRun` and `TaskRun` data off cluster, thereby ensuring data
+  is archived before removal.
+- Direct Pipelines as Code to use a URL pattern that displays the `PipelineRun`
+  or `TaskRun` info from the Hybrid Application Console (HAC). This ensures
+  end users do not need access to the underlying compute cluster(s).
+- The Pipelines as Code application name must match the GitHub Application name.
```
[the remark this answers is not in the document]
Comment (Roming22) [reply ADR/0009-pipeline-service-via-operator.md:39]: Updated

File: ADR/0009-pipeline-service-via-operator.md (hunk 4 above)
Comment (Roming22) [reply ADR/0009-pipeline-service-via-operator.md:40]: Updated

**Review by sbose78 — APPROVED** [review]
Looks good to me.

File: ADR/0009-pipeline-service-via-operator.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+- Direct Pipelines as Code to use a URL pattern that displays the `PipelineRun`
+  or `TaskRun` info from the Hybrid Application Console (HAC). This ensures
+  end users do not need access to the underlying compute cluster(s).
+- The Pipelines as Code application name must match the GitHub Application name, so that users understand which GitHubApplication is responsible for triggering the pipelines.
+- The GitHub Application secret value, deployed using an ExternalSecret.
+- Any configuration related to performance.
+
+Furthermore, as the service will be accessed through CodeReadyToolchain (CRT), the
+following changes are also specific to RHTAP:
+- Deploying a proxy (known as `SprayProxy`) on the CRT host cluster that redirects
```
Comment (chmouel) [inline ADR/0009-pipeline-service-via-operator.md:50]: (nit) maybe not part of this PR but it would be nice there was a document on how the redirection mechanism works, what is the reasoning behind it and if there is further plans to improve this..

**Comment by chmouel** [comment]
I can only speak for Pipelines-as-Code and this looks good to me...

/cc @concaf @vdemeester 

(for service first osp operator part)

File: ADR/0009-pipeline-service-via-operator.md [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+All Tekton APIs will be provided using the stock OpenShift Pipelines operator (OSP).
+In the spirit of developing in a "Service First" manner, RHTAP will deploy
+a candidate "nightly" release of the operator. The service will be defined in
+the [pipeline-service](https://github.com/openshift/pipeline-service)
+repository, which is then imported into
+[infra-deployments](https://github.com/redhat-appstudio/infra-deployments) as
+an ArgoCD application.
+
+Not all metrics required for operating the service are exposed natively by the
```
Comment (vdemeester) [inline ADR/0009-pipeline-service-via-operator.md:27]: See https://github.com/tektoncd/experimental/pull/953

Comment (gabemontero) [reply ADR/0009-pipeline-service-via-operator.md:27]: @vdemeester I was going to add https://github.com/tektoncd/experimental/pull/953 to the huddle agenda for Oct 11 but it looks like we have an all hands meeting at the same time.

but from what I've seen so far, theoretically I could see us transitioning at least some of the metrics we currently maintain in our "exporter" to that if it got promoted out of experimental, but I think there is a fair bit of logistics and roadmapping we need to do.

At least for this iteration of the ADR, are you OK @vdemeester (are are you OK @Roming22 ) if we add a bullet stating we are monitoring the experimental project?

I could any of these following worlds
- we transition everything in exporter to that new operator
- we live with exporter at least for some time
- we live in a world where we want both the operator,  but also the exporter for those metrics we think we need, but get too much pushback
- I'm also curious about how our associated grafana dashboards will get consumed wrt either upstream on downstream, assuming there is agreement that our dashboards would provide value outside of our RHTAP service specific use cases ... which if I had to bet, I bet pipeline users would find them useful.

File: ADR/0009-pipeline-service-via-operator.md (hunk 14 above)
Comment (Roming22) [reply ADR/0009-pipeline-service-via-operator.md:50]: I've opened a story so we create a ADR (or update this one) to include more information about the SprayProxy.

File: ADR/0009-pipeline-service-via-operator.md (hunk 15 above)
Comment (adambkaplan) [reply ADR/0009-pipeline-service-via-operator.md:27]: I agree with Gabe - we can keep our metrics exporter for now, but work upstream to bring our findings to the experimental controller.

File: ADR/0009-pipeline-service-via-operator.md (hunk 3 above)
Comment (adambkaplan) [reply ADR/0009-pipeline-service-via-operator.md:37]: I agree with Romain - ADR is structured as "this is what we intend to do, here is the why and what is impacted."

**Review by adambkaplan — APPROVED** [review]
LGTM
